ludwig/models/trainer.py

```diff
@@ -91,187 +101,209 @@ def __exit__(self, exc_type, exc_val, exc_tb):
         self.shutdown()
 
 
+@dataclass
+class TrainerConfig:
+    """TrainerConfig is a dataclass that configures most of the hyperparameters used for model training."""
+
+    optimizer: Optional[BaseOptimizer] = OptimizerDataclassField(default={"type": "adam"})
+    """Instance of `ludwig.modules.optimization_modules.BaseOptimizer` that specifies a torch-supported optimizer and
+       its attributes (default: `ludwig.modules.optimization_modules.AdamOptimizer()`)."""
+
+    epochs: int = schema.PositiveInteger(default=100)
+    "Number of epochs the algorithm is intended to be run over (default: 100)."
+
+    regularization_lambda: float = schema.FloatRange(default=0.0, min=0)
+    "Strength of the $L2$ regularization (default: 0.0)."
+
+    regularization_type: Optional[str] = schema.RegularizerOptions()
+    "Type of regularization, one of ('l1', 'l2', 'l1_l2', None) (default: 'l2')."
+
+    should_shuffle: bool = True
+    "Whether to shuffle batches during training when true (default: True)."
+
+    learning_rate: float = schema.NumericOrStringOptionsField(
+        default=0.001, default_numeric=0.001, min=0.0, max=1.0, options=["auto"], default_option="auto", nullable=False
+    )
+    """Learning rate specified in configuration, represents how much to scale the gradients by. If 'auto',
+       `tune_learning_rate` must be called before training to estimate the optimal learning rate. (default: 0.001)."""
+
+    batch_size: Union[int, str] = schema.IntegerOrStringOptionsField(
+        default=128, options=["auto"], nullable=False, min_exclusive=0
+    )
+    "Size of batch to pass to the model for training (default: 128)."
+
+    eval_batch_size: Union[None, int, str] = schema.IntegerOrStringOptionsField(
```

Comment:
This is an interesting case wrt. setting defaults in the schema (it's causing the latest error), would be interested in your thoughts @tgaddair @justinxzhao.

As the field is coded now, this is what is generated as the schema:
```
{
   "oneOf":[
      {
         "type":"integer",
         "exclusiveMinimum":0
      },
      {
         "type":"string",
         "enum":[
            "auto"
         ]
      },
      {
         "type":"null"
      }
   ]
}
```
With a `oneOf` construction across different types (e.g. integer and string-enum), there is no way to set a single "default" - you have to specify defaults for each option/sub-schema and there is no way in native JSON to set a macro level choice as to what the default value of the field should be. We are fine because marshmallow handles this so e.g. trying to make a TrainerConfig() object without specifying `eval_batch_size` will correctly create an instance with that attr. set to `None`.

But should the sub-schema have defaults at all? They would more or less just be metadata for the user.

Reply:
out of date